File: impl/typescript/src/topic/message-delivery.ts
/**
 * MessageDelivery -- wrapper returned by TopicAdapter.subscribe() for each message.
 *
 * Spec reference: 004 spec.md FR-T24, FR-T25, FR-T29
 *   - FR-T24: MessageDelivery wraps TopicMessage with backend metadata.
 *   - FR-T25: BackendSequence enables at-least-once delivery deduplication.
 *   - FR-T29: ConsensusTimestamp mapping per backend (authoritative, not sender-reported).
 *
 * Spec reference: 004 data-model.md MessageDelivery entity.
 *
 * Immutable value type.
 */

import type { TopicMessage } from './message.js';

/**
 * A received message with backend-assigned metadata.
 *
 * FR-T24: Each message delivered via Subscribe() is wrapped in a MessageDelivery
 * that provides the authoritative backend consensus timestamp and sequence number.
 *
 * FR-T29: The consensusTimestamp is assigned by the backend, NOT the sender's
 * self-reported timestamp. The adapter MUST NOT use the sender's timestamp.
 *
 * FR-T25: Consumers use backendSequence for at-least-once delivery deduplication.
 */
export interface MessageDelivery {
  /**
   * The deserialized TopicMessage envelope.
   * FR-T24: The original signed message as received from the topic.
   */
  readonly message: TopicMessage;

  /**
   * Authoritative backend clock when the message was finalized (nanoseconds).
   * FR-T24, FR-T29: Source depends on backend:
   * - HCS: Hedera consensus timestamp
   * - ERC-log: block.timestamp (converted to nanoseconds)
   * - Kafka: Anchor timestamp from HCS anchoring proof
   */
  readonly consensusTimestamp: bigint;

  /**
   * Backend-native ordering number.
   * FR-T24, FR-T25: Used for resumption and deduplication:
   * - HCS: topic sequence number
   * - ERC-log: blockNumber * 10000 + logIndex
   * - Kafka: partition offset
   */
  readonly backendSequence: bigint;
}

File: impl/golang/internal/topic/message_delivery.go
package topic

// FR-T24: subscribe() returns MessageDelivery
// MessageDelivery wraps a TopicMessage with backend-provided delivery metadata.
// It is the type received by subscribers when a message arrives from the transport.
//
// Fields follow FR-T24 (delivery envelope) and FR-T29 (backend sequence mapping):
//   - Message: The original TopicMessage as published
//   - ConsensusTimestamp: Backend-assigned consensus timestamp in Unix nanoseconds
//   - BackendSequence: Backend-assigned monotonic sequence number for ordering
type MessageDelivery struct {
	// Message is the delivered TopicMessage envelope.
	Message TopicMessage `json:"message"`
	// FR-T29: ConsensusTimestamp from backend authoritative clock
	// ConsensusTimestamp is the backend-assigned consensus timestamp in Unix nanoseconds.
	// For HCS this is the Hedera consensus timestamp. For Kafka this is the broker timestamp.
	ConsensusTimestamp uint64 `json:"consensusTimestamp"`
	// BackendSequence is the backend-assigned sequence number for total ordering.
	// For HCS this is the topic sequence number. For Kafka this is the partition offset.
	BackendSequence uint64 `json:"backendSequence"`
}
